src/domain_decomposition.hpp

```diff
@@ -13,97 +16,91 @@
-        // group until either the target group size is reached, or a cell with a
-        // different kind is encountered.
-        // On completion, cell_starts_ partitions the local gid into cell
-        // groups, and group_kinds_ records the cell kind in each cell group.
-        if (num_local_cells()>0) {
-            cell_size_type group_size = 1;
-
-            // 1st group starts at cell_begin_
-            group_starts_.push_back(cell_begin_);
-            auto group_kind = rec.get_cell_kind(cell_begin_);
-
-            // set kind for 1st group
-            group_kinds_.push_back(group_kind);
-
-            cell_gid_type gid = cell_begin_+1;
-            while (gid<cell_end_) {
-                auto kind = rec.get_cell_kind(gid);
-
-                // Test if gid belongs to a new cell group, i.e. whether it has
-                // a new cell_kind or if the target group size has been reached.
-                if (kind!=group_kind || group_size>=rules.target_group_size) {
-                    group_starts_.push_back(gid);
-                    group_kinds_.push_back(kind);
-                    group_size = 0;
-                }
-                ++group_size;
-                ++gid;
-            }
-            group_starts_.push_back(cell_end_);
+        auto b = first_cell(domain_id_);
+        auto e = first_cell(domain_id_+1);
+
+        // Local load balance
+
+        std::unordered_map<kind_type, std::vector<cell_gid_type>> kind_lists;
+        for (auto gid: make_span(b, e)) {
+            kind_lists[rec.get_cell_kind(gid)].push_back(gid);
         }
-    }
 
-    /// Returns the local index of the cell_group that contains a cell with
-    /// with gid.
-    /// If the cell is not on the local domain, the optional return value is
-    /// not set.
-    util::optional<cell_size_type>
-    local_group_from_gid(cell_gid_type gid) const {
-        // check if gid is a local cell
-        if (!is_local_gid(gid)) {
-            return util::nothing;
+        // Create a flat vector of the cell kinds present on this node,
+        // partitioned such that kinds for which GPU implementation are
+        // listed before the others. This is a very primitive attempt at
+        // scheduling; the cell_groups that run on the GPU will be executed
+        // before other cell_groups, which is likely to be more efficient.
+        std::vector<cell_kind> kinds;
+        for (auto l: kind_lists) {
+            kinds.push_back(cell_kind(l.first));
         }
-        return gid_group_partition().index(gid);
-    }
+        std::partition(kinds.begin(), kinds.end(), has_gpu_backend);
 
-    /// Returns the gid of the first cell on the local domain.
-    cell_gid_type cell_begin() const {
-        return cell_begin_;
+        for (auto k: kinds) {
+            // put all cells into a single cell group on the gpu if possible
+            if (node_.num_gpus && has_gpu_backend(k)) {
```

Comment:
You bring up a good point.

This is the only point where the decision about which back end to use for cell groups is made. It uses the simple method:

1. if a gpu is available and if the cell_kind has a gpu implementation, create a single cell_group with all cells of cell_kind to run on on the gpu
2. else create many cell groups, with 1 cell per group to run on the cpu.

This is not great. For one, as @halfflat points out, it does not consider the relative benefits of placing the work on cpu or gpu, group sizes, and so on.

This patch is intended as the smallest possible change that puts these decisions "in the right place", even if the actual decision making logic is too simple.

There will be a follow up patch that will split the load balancing and domain decomposition descriptions. Then we can focus on the load balancer that will make more general decisions based on cost models etc.